### ISSUE TITLE
Pandas KeyError in _validate_read_indexer

### ISSUE THREAD

**Issue by palbanese** [debbiemarkslab/EVcouplings#246]
KeyError                                  Traceback (most recent call last)
<ipython-input-21-b280763115a0> in <module>
      3 
      4 config = read_config_file("/home/pascal/evcouplings/Config_files/PsbS_X_config_test.txt")
----> 5 outcfg = execute(**config)

~/anaconda3/lib/python3.7/site-packages/evcouplings/utils/pipeline.py in execute(**config)
    174 
    175             # run stage
--> 176             outcfg = runner(**incfg)
    177 
    178             # prefix output keys if this parameter is

~/anaconda3/lib/python3.7/site-packages/evcouplings/align/protocol.py in run(**kwargs)
   1671         )
   1672 
-> 1673     return PROTOCOLS[kwargs["protocol"]](**kwargs)

~/anaconda3/lib/python3.7/site-packages/evcouplings/align/protocol.py in complex(**kwargs)
   1567         )
   1568 
-> 1569     outcfg = PROTOCOLS[kwargs["alignment_protocol"]](**kwargs)
   1570 
   1571     # if the user selected the existing alignment protocol

~/anaconda3/lib/python3.7/site-packages/evcouplings/align/protocol.py in standard(**kwargs)
   1483     if kwargs["extract_annotation"]:
   1484         annotation_file = prefix + "_annotation.csv"
-> 1485         annotation = extract_header_annotation(ali_raw)
   1486         annotation.to_csv(annotation_file, index=False)
   1487 

~/anaconda3/lib/python3.7/site-packages/evcouplings/align/protocol.py in extract_header_annotation(alignment, from_annotation)
    463 
    464     df = pd.DataFrame(res)
--> 465     return df.loc[:, ["id", "name"] + list(col_to_descr.keys())]
    466 
    467 

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py in __getitem__(self, key)
   1759                 except (KeyError, IndexError, AttributeError):
   1760                     pass
-> 1761             return self._getitem_tuple(key)
   1762         else:
   1763             # we by definition only have the 0th axis

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py in _getitem_tuple(self, tup)
   1286                 continue
   1287 
-> 1288             retval = getattr(retval, self.name)._getitem_axis(key, axis=i)
   1289 
   1290         return retval

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py in _getitem_axis(self, key, axis)
   1951                     raise ValueError("Cannot index with multidimensional key")
   1952 
-> 1953                 return self._getitem_iterable(key, axis=axis)
   1954 
   1955             # nested tuple slicing

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py in _getitem_iterable(self, key, axis)
   1592         else:
   1593             # A collection of keys
-> 1594             keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
   1595             return self.obj._reindex_with_indexers(
   1596                 {axis: [keyarr, indexer]}, copy=True, allow_dups=True

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py in _get_listlike_indexer(self, key, axis, raise_missing)
   1550 
   1551         self._validate_read_indexer(
-> 1552             keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
   1553         )
   1554         return keyarr, indexer

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py in _validate_read_indexer(self, key, indexer, axis, raise_missing)
   1652             if not (ax.is_categorical() or ax.is_interval()):
   1653                 raise KeyError(
-> 1654                     "Passing list-likes to .loc or [] with any missing labels "
   1655                     "is no longer supported, see "
   1656                     "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"  # noqa:E501

KeyError: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'

**Comment by palbanese** [comment]
Hi, i'm new here (also to python code). But this seem something with the syntax of the tables?
"Passing list-likes to .loc or [] with any missing labels is no longer supported"

**Comment by palbanese** [comment]
Is this about the release of Pandas?

**Comment by palbanese** [comment]
we also tried to remove anaconda and install directly from python, changing nearly all input files and reinstall the external tools. Still getting the same error out as above. Any help would be highly appreciated.

The environment is  Ubuntu 20.04.1 LTS (GNU/Linux 5.4.0-42-generic x86_64)

**Comment by thomashopf** [comment]
Hi @palbanese, this is a known problem where we used outdated pandas syntax for too long (see #231). 

The issue has already been fixed, but not yet released to pypi (i.e. the version of the package that you get using pip install evcouplings). 

For now, the solution is to install the develop branch version of the package. The easiest way to do so is to download https://github.com/debbiemarkslab/EVcouplings/archive/develop.zip and then `pip install develop.zip`

Does that fix the issue for you?

**Comment by palbanese** [comment]
Hi thomashopf, 
thanks fro your help, and yes, it worked out nicely now!.. maybe it's a good idea to put just a note on the README file with the installation instruction, you'll avoid this kind of issues reported in the near future :) 

I encountered another issue btw, I don't know if I should open another thread:

It concerns the compare protocol, apparently when it compiles the SIFTS mapping tables there are connection issues, because the csv was there, but only the PDB accessions starting with 1 and some with 2. So i taught was really a connection issue, and i deleted the partial table and restareted the job with another name so see if the tables would have been completed, but same error again, and the table was compiled to the same exact point. Is it the UNIPROT server limiting the number of access? is i possible to get these table from another source, like https://www.ebi.ac.uk/pdbe/docs/sifts/quick.html or available from your repository?

here is the log:
ResourceError                             Traceback (most recent call last)
In  [10]:
Line 5:     outcfg = execute(**config)

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/utils/pipeline.py, in execute:
Line 187:   outcfg = runner(**incfg)

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/protocol.py, in run:
Line 1250:  return PROTOCOLS[kwargs["protocol"]](**kwargs)

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/protocol.py, in complex:
Line 961:   outcfg, first_sifts_map = _identify_monomer_structures("first", outcfg, first_aux_prefix)

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/protocol.py, in _identify_monomer_structures:
Line 945:   sifts_map, sifts_map_full = _identify_structures(

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/protocol.py, in _identify_structures:
Line 152:   s = SIFTS(

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/sifts.py, in __init__:
Line 284:   self.create_sequence_file(sequence_file)

File /home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/sifts.py, in create_sequence_file:
Line 456:   raise ResourceError(

ResourceError: Could not fetch sequences for SIFTS mapping tables from UniProt since maximum number of retries after connection errors was exceeded. Retry at a later time, or call SIFTS.create_sequence_file() with a higher value for max_retries.

**Comment by palbanese** [comment]
small update:
it seems that this is caused by the uniprot server, even if i raise the number of max_retries to 1000, i get this error out: 

SSLError: HTTPSConnectionPool(host='www.uniprot.org', port=443): Max retries exceeded with url: /mapping/ (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'ssl3_read_bytes', 'sslv3 alert handshake failure')])")))

I tried also to reduce the number of sequences to fetch down to 100/round and the .csv table is compiled for 100 entries, therefore this is apparently an issue of the loop function which is not appreciated by the host server. 

Any suggestions? was this issue already reported?

**Comment by thomashopf** [comment]
Hi @palbanese,

Sorry for the inconvenience with the installation, we urgently need to put out a new public release but there are a few considerations and more testing involved before we can do so.

Re the mapping files - the sequence download from Uniprot is unfortunately notoriously unreliable (which is why I added the retries option at some point), and becomes more so when putting in higher numbers of entries. One day it works, the next day it just breaks...

Ultimately, we should precompute the files for our users and fetch them from our own servers instead.

In the meantime, you can download the files here:
https://www.dropbox.com/s/lltxjvzfj3kx552/sifts_2020-06-20.csv?dl=0
https://www.dropbox.com/s/dsudce1lfaus24f/sifts_2020-06-20.fasta?dl=0

**Comment by palbanese** [comment]
Thanks for the prompt support! 

Finally, all outputs are there, amazing! :)

I think only one small last step is going wrong in preparing the output archive folder, seems that some color/fonts are missing from matplotlib package? or these ['dtype=object' when creating the ndarray]?


see log:
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/parameters.py:137: RuntimeWarning: divide by zero encountered in log2
  B = -fi * np.log2(fi)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/parameters.py:137: RuntimeWarning: invalid value encountered in multiply
  B = -fi * np.log2(fi)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:458: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  return np.array(
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/distances.py:701: RuntimeWarning: All-NaN slice encountered
  agg_mat = agg_func(new_mat, axis=0)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/distances.py:701: RuntimeWarning: All-NaN slice encountered
  agg_mat = agg_func(new_mat, axis=0)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:458: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  return np.array(
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:458: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  return np.array(
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:458: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  return np.array(
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/compare/pdb.py:588: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  np.array([
findfont: Font family ['Arial'] not found. Falling back to DejaVu Sans.
findfont: Font family ['Arial'] not found. Falling back to DejaVu Sans.
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/mutations.py:505: MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("RdBu_r"))
  colormap.set_bad(na_color)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/mutations.py:507: MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("Oranges"))
  colormap_conservation.set_bad(na_color)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/mutations.py:505: MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("RdBu_r"))
  colormap.set_bad(na_color)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/mutations.py:507: MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("Oranges"))
  colormap_conservation.set_bad(na_color)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/mutations.py:505: MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("RdBu_r"))
  colormap.set_bad(na_color)
/home/pascal/.local/lib/python3.8/site-packages/evcouplings/visualize/mutations.py:507: MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("Oranges"))
  colormap_conservation.set_bad(na_color)
​

**Comment by thomashopf** [comment]
@palbanese Great!

The rest are just warnings, so no need to worry about these - but I will make a note to fix the deprecations so we don't run into problems in the future.

I am closing this issue but feel free to reopen if you encounter any more issues around these.
